aten/src/ATen/native/cuda/Unique.cu

```diff
@@ -15,86 +16,103 @@ namespace native{
-        thrust::scatter(policy, inv_loc_ptr, inv_loc_ptr + num_inp, sorted_indices_ptr, inverse_indices_ptr);
-        inverse_indices.resize_(input.sizes());
-    }
 
-    // unique
-    Tensor counts = at::empty({0}, self.options().dtype(kLong));
-    if (!return_counts) {
-      int64_t num_out = thrust::unique(policy, output_data, output_data + num_inp) - output_data;
-      output.resize_(num_out);
-    } else {
-      Tensor sorted_indices = at::arange(0, num_inp + 1, self.type().toScalarType(kLong));
-      int64_t* sorted_indices_ptr = sorted_indices.data<int64_t>();
-      int64_t num_out = thrust::unique_by_key(policy, output_data, output_data + num_inp, sorted_indices_ptr).first - output_data;
-      sorted_indices[num_out] = num_inp;
-      output.resize_(num_out);
-      counts.resize_(num_out);
-      int64_t* counts_ptr = counts.data<int64_t>();
-      thrust::adjacent_difference(policy, sorted_indices_ptr + 1, sorted_indices_ptr + num_out + 1, counts_ptr);
+    Tensor inverse_indices, counts;
+    int64_t num_out;
+    std::tie(inverse_indices, counts, num_out) = compute_unique(output_data, num_inp, return_inverse, return_counts, self.options().dtype(kLong));
+    output.resize_(num_out);
+
+    if (return_inverse) {
+        inverse_indices.resize_(self.sizes());
     }
 
-    THCudaCheck(cudaGetLastError());
     return std::tuple<Tensor, Tensor, Tensor>(output, inverse_indices, counts);
-  }
+}
 
 template <typename scalar_t>
-  std::tuple<Tensor, Tensor, Tensor> _unique_dim_cuda_template(
+std::tuple<Tensor, Tensor, Tensor> unique_dim_cuda_template(
     const Tensor& self,
     const int64_t dim,
     const bool return_inverse,
     const bool return_counts) {
 
-    cudaStream_t stream = at::cuda::getCurrentCUDAStream();
-    auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
-    auto policy = thrust::cuda::par(allocator).on(stream);
-
-    Tensor input_flat = self.transpose(dim, 0);
-    auto orig_sizes = input_flat.sizes().vec();
-    input_flat = input_flat.contiguous().view({input_flat.size(0), -1});
-
-    scalar_t* input_flat_ptr = input_flat.data<scalar_t>();
-
-    Tensor indices = at::arange(0, input_flat.size(0), self.options().dtype(kLong));
-    int64_t* indices_ptr = indices.data<int64_t>();
+    int64_t num_inp = self.size(0);
+    Tensor input_flat = self.transpose(dim, 0).contiguous().view({num_inp, -1});
     int64_t numel = input_flat.size(1);
+    scalar_t *input_flat_ptr = input_flat.data<scalar_t>();
 
-    // sort indices using data
-    thrust::sort(policy, indices_ptr, indices_ptr + indices.numel(),
+    Tensor indices = at::arange(0, num_inp, self.options().dtype(kLong));
+    int64_t *indices_data = indices.data<int64_t>();
+
+    Tensor inverse_indices, counts;
+    int64_t num_out;
+    std::tie(inverse_indices, counts, num_out) = compute_unique(
+      indices_data, num_inp, return_inverse, return_counts, self.options().dtype(kLong),
```

Comment:
compute_unique first argument type is `scalar_t *`, and you are sending `indices_data` which is `int64_t *`, does not look right.

Reply:
@ngimel This is expected. The tensor passed in is the index tensor, and we override the compare operator of it to compare the actual underlying data, and finally use an `index_select` to get the result.

Reply:
So `scalar_t` is resolved to `int64_t`. Confusing, but ok. However, I don't think this line https://github.com/pytorch/pytorch/pull/18459/files#diff-0ac6387c4c1640433eaf96ce504cb53bR50 works as expected if `data` is in fact indices - the comparison operator has to be overriden here too.

Reply:
@ngimel I think you are right. There is another problem that the test case is not covering cuda, so we didn't see any error on it. I have fixed the test. But not the implementation itself yet.

Reply:
@ngimel implementation also fixed now